components/api/src/main/java/com/hotels/styx/api/service/spi/AbstractStyxService.java

```diff
@@ -0,0 +1,107 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hotels.styx.api.service.spi;
+
+import com.google.common.collect.ImmutableMap;
+import com.hotels.styx.api.HttpHandler;
+import com.hotels.styx.api.HttpRequest;
+import com.hotels.styx.api.HttpResponse;
+import rx.Observable;
+
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+import static com.hotels.styx.api.service.spi.StyxServiceStatus.CREATED;
+import static com.hotels.styx.api.service.spi.StyxServiceStatus.FAILED;
+import static com.hotels.styx.api.service.spi.StyxServiceStatus.RUNNING;
+import static com.hotels.styx.api.service.spi.StyxServiceStatus.STARTING;
+import static com.hotels.styx.api.service.spi.StyxServiceStatus.STOPPED;
+import static com.hotels.styx.api.service.spi.StyxServiceStatus.STOPPING;
+import static io.netty.handler.codec.http.HttpResponseStatus.OK;
+import static java.lang.String.format;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static rx.Observable.just;
+
+/**
+ * A helper class for implementing StyxService interface.
+ *
+ * AbstractStyxService provides service state management facilities
+ * for implementing a StyxSerive interface.
+ */
+public abstract class AbstractStyxService implements StyxService {
+    private final String name;
+    private final AtomicReference<StyxServiceStatus> status = new AtomicReference<>(CREATED);
+
+    public AbstractStyxService(String name) {
+        this.name = name;
+    }
+
+    public StyxServiceStatus status() {
+        return status.get();
+    }
+
+    protected CompletableFuture<Void> startService() {
+        return completedFuture(null);
+    }
+
+    protected CompletableFuture<Void> stopService() {
+        return completedFuture(null);
+    }
+
+    @Override
+    public CompletableFuture<Void> start() {
+        boolean changed = status.compareAndSet(CREATED, STARTING);
+
+        if (changed) {
+            return startService()
+                    .exceptionally(failWithMessage("Service failed to start."))
+                    .thenAccept(na -> status.compareAndSet(STARTING, RUNNING));
+        } else {
+            throw new IllegalStateException(format("Start called in %s state", status.get()));
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> stop() {
+        boolean changed = status.compareAndSet(RUNNING, STOPPING);
```

Comment:
I'm just thinking, if does it make sense to signalize that someone is trying to stop service that is not running?